src/common/TrackedOp.h

```diff
@@ -233,28 +233,124 @@ class TrackedOp : public boost::intrusive::list_base_hook<> {
   utime_t initiated_at;
 
   struct Event {
-    utime_t stamp;
-    std::string str;
+    utime_t stamp_;
+    string str_;
+    const char *cstr_ = nullptr;
+
+    Event() {};
+    Event(utime_t t, std::string_view s) : stamp_(t), str_(s) {}
+    Event(utime_t t, const char *s) : stamp_(t), cstr_(s) {}
+
+    void set(utime_t t, std::string_view s) {
+      stamp_ = t;
+      str_ = s;
+    }
 
-    Event(utime_t t, std::string_view s) : stamp(t), str(s) {}
+    void set(utime_t t, const char* s) {
+      stamp_ = t;
+      cstr_ = s;
+    }
+
+    const utime_t stamp() const {
+      return stamp_;
+    }
 
-    int compare(const char *s) const {
-      return str.compare(s);
+    const int compare(const char *s) const {
+      if (cstr_)
+	return strcmp(cstr_, s);
+      else
+	return str_.compare(s);
     }
 
     const char *c_str() const {
-      return str.c_str();
+      if (cstr_)
+	return cstr_;
+      else
+	return str_.c_str();
     }
 
     void dump(ceph::Formatter *f) const {
-      f->dump_stream("time") << stamp;
-      f->dump_string("event", str);
+      f->dump_stream("time") << stamp_;
+      f->dump_string("event", c_str());
+    }
+  };
+
+  // Semi-Lockless append only list with reserved items
+  template<class T>
+  class AppendList {
+    static constexpr size_t RESERVED_ITEMS = 16;
+
+  private:
+    // pre-initialized array with atomic index
+    std::array<T, RESERVED_ITEMS> arr;
+    std::atomic<size_t> reservation_ = 0;
+    std::atomic<size_t> size_ = 0;
+    // Expansion storage with lock
+    std::vector<T> vec;
+    mutable ceph::mutex lock = ceph::make_mutex("AppendList::lock");
```

Comment:
Why not `boost::small_vector`?

Reply:
I think the tricky part would be figuring out how to avoid a global lock over the whole data structure.  The append-only nature of this list is why we can get away with using the two atomic counters for the std:array portion and then drop back to the lock for the vector.  Since we (almost?) never actually use the vector though since the number of events is typically low, we rarely actually invoke the lock.

Reply:
Alternately, we may want to consider something like this (and maybe more generally in other places):

https://github.com/cameron314/concurrentqueue

The tradeoffs are going to take a bit of thought though (see reasons not to use).

Also a potentially very interesting paper mentioned in the readme:

http://webee.technion.ac.il/~idish/ftp/spaa049-gidron.pdf

Reply:
I'm just really skeptical about the advantages of these tricks and it adds a lot of (mental) complexity to the code. At least this should be abstracted out as another data structure so that its interface and constraints are clear/documented.

Reply:
@markhpc is this AppendList new? Maybe you already did that and I apologize.

Reply:
cohort had (archive) a single lock-free queue apparently similar to concurrentqueue (based onhttps://github.com/mstump/queues)--for the actual op queue.  As I've noted before, we hit the numa issue--but we could still get something like 2m synthetic ops through it on a single i7.  We discarded optracker.